cumulusci/utils/fileutils.py

```diff
@@ -86,6 +92,206 @@ def load_from_source(
+        path = fspath.join(self.filename, other)
+        return FSResource.new(self.fs.geturl(path))
+
+    def copy_to(self, other):
+        """Create a new FSResource by copying the underlying resource
+
+        Note that calling .close() on either one (or exiting the
+        context of the original) will close the filesystem that both use.
+
+        In practice, if you use the new one within the open context
+        of the old one, you'll be fine.
+        """
+        if isinstance(other, (str, Path)):
+            other = FSResource.new(other)
+        copy.copy_file(self.fs, self.filename, other.fs, other.filename)
+
+    def mkdir(self, *, parents=False, exist_ok=False):
+        if parents:
+            self.fs.makedirs(self.filename, recreate=exist_ok)
+        else:
+            self.fs.makedir(self.filename, recreate=exist_ok)
+
+    def __contains__(self, other):
+        return other in str(self.geturl())
+
+    def __truediv__(self, other):
+        return self.joinpath(other)
+
+    def __repr__(self):
+        return f"<FSResource {self.geturl()}>"
+
+    def __str__(self):
+        rc = self.geturl()
+        if rc.startswith("file://"):
+            return rc[6:]
+
+    def __fspath__(self):
+        return self.fs.getsyspath(self.filename)
+
+    def close(self):
+        self.fs.close()
+
+    @staticmethod
+    @contextmanager
+    def open_fs_resource(
+        resource_url_or_path: Union[str, Path, "FSResource"], filesystem: base.FS = None
+    ):
+        """Create a context-managed FSResource
+
+        Input is a URL, path (absolute or relative) or FSResource
+
+        The function should be used in a context manager. The
+        resource's underlying filesystem will be closed automatically
+        when the context ends and the data will be saved back to the
+        filesystem (local, remote, zipfile, etc.)
+
+        Think of it as a way of "mounting" a filesystem, directory or file.
+
+        For example:
+
+        >>> from tempfile import TemporaryDirectory
+        >>> with TemporaryDirectory() as tempdir:
+        ...     abspath = Path(tempdir) / "blah"
+        ...     with open_fs_resource(abspath) as fs:
+        ...         fs.mkdir()
+        ...     newfile = fs / "newfile"
+        ...     with newfile.open("w") as f:
+        ...         _ = f.write("xyzzy")
+        ...     with newfile.open("r") as f:
+        ...         print(f.read())
+        xyzzy
+
+        >>> with open_fs_resource("cumulusci.yml") as cumulusci_yml:
+        ...      with cumulusci_yml.open() as c:
+        ...          print(c.read(5))
+        proje
+
+        """
+        resource = FSResource.new(resource_url_or_path, filesystem)
+        if not filesystem:
+            filesystem = resource
+        try:
+            yield resource
+        finally:
+            filesystem.close()
+
+
+open_fs_resource = FSResource.open_fs_resource
+
 if __name__ == "__main__":  # pragma: no cover
     import doctest
 
```

Comment:
Do the doctests also get discovered and run by pytest?

Reply:
Not automatically but I have a method called `test_docstrings` in `cumulusci/utils/tests/test_fileutils.py` which does it (for this module).